### PR TITLE
[Review] PD Interfaces

### PR DESCRIPTION
**Pull request by quasiben**
This PR is an exploration of what it would take to make https://github.com/rapidsai/dask-cudf/pull/122 functional

### CODE REVIEW

File: python/cudf/dataframe/dataframe.py [hunk 1]
```diff
@@ -2344,6 +2331,35 @@ def quantile(self,
...
+
+        for x in self._cols.values():
+            if str(x.dtype) in include:
+                df.add_column(x.name, x.data)
+
+        return df
+
+    @property
+    def ndim(self):
+        return len(self.columns)
```
Comment (kkraus14) [inline python/cudf/dataframe/dataframe.py]: I believe this is supposed to only return 2 for dataframes and 1 for series. See http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.ndim.html

File: python/cudf/dataframe/dataframe.py [hunk 2]
```diff
@@ -2383,54 +2399,24 @@ def __init__(self, df):
...
-            else:
-                raise TypeError(type(arg_1))
-            col_slice = arg_2
-
+        if isinstance(arg, tuple):
+            row_slice, col_slice = arg
+            if isinstance(col_slice, pd.Series):
+                # equivalent of core.indexing._getitem_iterable
+                df = pd.DataFrame(index=self._df.index)
+                return df
```
Comment (kkraus14) [inline python/cudf/dataframe/dataframe.py]: I'm not following the logic here. You're returning an empty pandas DataFrame with the index without using the input tuple. Could you explain this?

File: python/cudf/datasets.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+def make_categorical(n, rstate):
+    return pd.Categorical.from_codes(rstate.randint(0, len(names), size=n),
+                                     names)
+
+
+make = {float: make_float,
+        int: make_int,
+        str: make_string,
+        object: make_string,
+        'category': make_categorical}
```
Comment (kkraus14) [inline python/cudf/datasets.py]: I think we should change this to allow taking in the types people would expect to use (numpy and pandas dtypes). As all of these inputs can be easily converted to numpy / pandas dtypes.

**Comment by mrocklin** [comment]
@quasiben this branch seems to also have some timeseries code in it.  Is this intentional?

**Comment by quasiben** [comment]
@mrocklin, yes, though I'm happy to move it.  I wanted an easy way to generate data for test the select_dtypes function

File: python/cudf/dataframe/dataframe.py (hunk 1 above)
Comment (quasiben) [reply python/cudf/dataframe/dataframe.py]: i'm going to remove `ndim` for now.  I like the way pandas does it by calling out to len(self.axes) and we don't currently have that.  I'll that build in a separate PR

File: python/cudf/dataframe/dataframe.py [hunk 4]
```diff
@@ -2344,6 +2362,35 @@ def quantile(self,
...
+        Parameters
+        ----------
+        include : [type]
+            [description] (the default is None, which [default_description])
+
+        """
+
+        # TODO: mechanism to hand int->int64 float->float64...
+
+        if isinstance(include, str):
```
Comment (kkraus14) [inline python/cudf/dataframe/dataframe.py]: What if `include` is a numpy / pandas dtype?

File: python/cudf/dataframe/dataframe.py (hunk 2 above)
Comment (quasiben) [reply python/cudf/dataframe/dataframe.py]: This should be gone now.  I was doing this while trying to trace some functionality around select_dtypes

File: python/cudf/dataframe/dataframe.py (hunk 4 above)
Comment (quasiben) [reply python/cudf/dataframe/dataframe.py]: That's a good question.  Last night I would looking around the codebase for unified handling of np.int64/int, float/np.float64.  Does such a thing exist ?  If not, I'll build it

Comment (quasiben) [reply python/cudf/dataframe/dataframe.py]: @kkraus14 what do you think about using `pd.core.dtypes.common.pandas_dtype` for checking/conversion ?

Comment (quasiben) [reply python/cudf/dataframe/dataframe.py]: nevermind, I'll just use `np.dtype()` should be fine

Comment (kkraus14) [reply python/cudf/dataframe/dataframe.py]: > nevermind, I'll just use `np.dtype()` should be fine

That won't handle the `category` dtype is the only problem.

Comment (kkraus14) [reply python/cudf/dataframe/dataframe.py]: I think you can use `pd.api.types.pandas_dtype(...)`

File: python/cudf/dataframe/dataframe.py [hunk 5]
```diff
@@ -2344,6 +2362,28 @@ def quantile(self,
...
+        include : [type]
+            [description] (the default is None, which [default_description])
+
+        """
+
+        if not isinstance(include, (list, tuple)):
+            include = [include]
+        df = DataFrame()
+
+        include = [np.dtype(d) for d in include]
```
Comment (kkraus14) [inline python/cudf/dataframe/dataframe.py]: This won't work with `categorical` dtype.

**Comment by quasiben** [comment]
@kkraus14 when you have a moment can you review again?

File: python/cudf/datasets.py (hunk 3 above)
Comment (kkraus14) [reply python/cudf/datasets.py]: @quasiben What are your thoughts on this one?

Comment (quasiben) [reply python/cudf/datasets.py]: Here I was following the convention which dask uses.   I like that I don't have to import numpy to define the dtypes and get up and running:

```
import cudf
gdf = cudf.datasets.timeseries(...)
```

File: python/cudf/dataframe/dataframe.py (hunk 4 above)
Comment (quasiben) [reply python/cudf/dataframe/dataframe.py]: Switched to `pd.api.types.pandas_dtype`

File: python/cudf/datasets.py (hunk 3 above)
Comment (quasiben) [reply python/cudf/datasets.py]: cool

**Pull request merged**